bdns/dns.go

```diff
@@ -475,3 +488,60 @@ func (dnsClient *DNSClientImpl) LookupCAA(ctx context.Context, hostname string)
 	}
 	return CAAs, nil
 }
+
+// logDNSError logs the provided err result from making a query for hostname. If
+// the err is a `dns.ErrId` instance then the raw Base64 URL encoded bytes of
+// the query (and if not-nil, the response) in wire format is logged as well.
+// This function is called from exchangeOne only for the case where an error
+// occurs querying a hostname that indicates a problem between the VA and the
+// resolver.
+func (dnsClient *DNSClientImpl) logDNSError(hostname string, msg, resp *dns.Msg, err error) {
+	// We use a stand-alone function for this to make it easy to share the
+	// implementation between the DNSClientImpl and MockDNSClient
+	logDNSError(dnsClient.log, hostname, msg, resp, err)
+}
+
+// log a DNS level error with the given logger. See DNSClientImpl.logDNSError.
+func logDNSError(logger blog.Logger, hostname string, msg, resp *dns.Msg, underlying error) {
+	// We don't expect logDNSError to be called with a nil msg or err but
+	// if it happens return early. We allow resp to be nil.
+	if msg == nil || len(msg.Question) == 0 || underlying == nil {
+		return
+	}
+	queryType := dns.TypeToString[msg.Question[0].Qtype]
+
+	// If the error indicates there was a query/response ID mismatch then we want
+	// to log more detail.
+	if underlying == dns.ErrId {
```

Comment:
Do you think there would be any value in adding a prom metric for this? It might be nice to be able to see if the occurrences of this correlate with any other host data we are collecting.

Reply:
Sure, I can see how that might end up being useful. Added in 4e4419f, labelled by chosen server and query type.